m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+        }
+
+        return metricTagSet;
+    }
+
+    private MetricTag createMetricTag(String tagName, String tagValue) {
+        MetricTag metricTag = resourcePool.getMetricTag();
+        metricTag.setTagName(tagName);
+
+        if (tagValue != null && !tagValue.isEmpty()) {
+            metricTag.setTagValue(tagValue);
+        }
+
+        return metricTag;
+    }
+
+    private int calculateSize(Metric metric) {
+        try {
+            synchronized (calcLock) {
+                metric.write(calcProtocol);
+
+                return calc.getSizeAndReset();
+            }
+        } catch (TException e) {
+            return DEFAULT_METRIC_SIZE;
+        }
+    }
+
+    private void reportCopyMetric(
+        Metric metric,
+        int size,
+        MetricType metricType,
+        long iValue,
+        double dValue
+    ) {
+        Metric metricCopy = resourcePool.getMetric();
+        metricCopy.setName(metric.getName());
+        metricCopy.setTags(metric.getTags());
+        // Unfortunately, there is no finer time resolution for Java 7
+        metricCopy.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metricCopy.setMetricValue(resourcePool.getMetricValue());
+
+        switch (metricType) {
+            case COUNTER:
+                CountValue countValue = resourcePool.getCountValue();
+                countValue.setI64Value(iValue);
+                metricCopy.getMetricValue().setCount(countValue);
+                break;
+            case GAUGE:
+                GaugeValue gaugeValue = resourcePool.getGaugeValue();
+                gaugeValue.setDValue(dValue);
+                metricCopy.getMetricValue().setGauge(gaugeValue);
+                break;
+            case TIMER:
+                TimerValue timerValue = resourcePool.getTimerValue();
+                timerValue.setI64Value(iValue);
+                metricCopy.getMetricValue().setTimer(timerValue);
+                break;
+        }
+
+        try {
+            metricQueue.put(new SizedMetric(metricCopy, size));
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String valueBucketString(double bucketBound) {
+        if (bucketBound == Double.MAX_VALUE) {
+            return "infinity";
+        }
+
+        if (bucketBound == -Double.MAX_VALUE) {
+            return "-infinity";
+        }
+
+        return String.format(bucketValFmt, bucketBound);
+    }
+
+    private String durationBucketString(Duration bucketBound) {
+        if (Duration.MAX_VALUE.equals(bucketBound)) {
+            return "infinity";
+        }
+
+        if (Duration.MIN_VALUE.equals(bucketBound)) {
+            return "-infinity";
+        }
+
+        return bucketBound.toString();
+    }
+
+    private class Processor implements Runnable {
+        @Override
+        public void run() {
+            try {
+                List<Metric> metrics = new ArrayList<>(freeBytes / 10);
```

Comment:
how'd you arrive at `freeBytes/10`?

Reply:
per our chat yesterday, seems like this number is just straight from the go implementation. 

Could you add an issue to profile the code for different workloads, to gauge expected perf, and to size the defaults/constants correctly. Might be worth applying for [YourKit](https://www.yourkit.com/)'s free OSS license